pkg/virt-controller/watch/drain/evacuation/evacuation.go

```diff
@@ -0,0 +1,493 @@
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		return nil
+	}
+
+	return vmi.(*virtv1.VirtualMachineInstance)
+}
+
+// Run runs the passed in NodeController.
+func (c *EvacuationController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+	log.Log.Info("Starting evacuation controller.")
+
+	// Wait for cache sync before we start the node controller
+	cache.WaitForCacheSync(stopCh, c.migrationInformer.HasSynced, c.vmiInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping evacuation controller.")
+}
+
+func (c *EvacuationController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *EvacuationController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VirtualMachineInstance %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VirtualMachineInstance %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *EvacuationController) execute(key string) error {
+
+	// Fetch the latest node state from cache
+	obj, exists, err := c.nodeInformer.GetStore().GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		c.migrationExpectations.DeleteExpectations(key)
+		return nil
+	}
+
+	if !c.migrationExpectations.SatisfiedExpectations(key) {
+		return nil
+	}
+
+	node := obj.(*k8sv1.Node)
+
+	// If the node has not taints, we have nothing to do
+	if !nodeHasTaint(node) {
+		return nil
+	}
+
+	vmis, err := c.listVMIsOnNode(node.Name)
+	if err != nil {
+		// XXX
+		return err
+	}
+
+	migrations, err := c.listRunningMigrations()
+
+	if err != nil {
+		// XXX
+		return err
+	}
+
+	return c.sync(node, vmis, migrations)
+
+}
+
+func (c *EvacuationController) sync(node *k8sv1.Node, vmisOnNode []*virtv1.VirtualMachineInstance, activeMigrations []*virtv1.VirtualMachineInstanceMigration) error {
+	if len(vmisOnNode) == 0 {
+		// nothing to do
+		return nil
+	}
+
+	if len(activeMigrations) >= 5 {
+		// Don't overload the cluster with migrations
+		return nil
+	}
+
+	now := time.Now()
+
+	notMigratingVMIs := c.filterRunningNonMigratingVMIs(vmisOnNode, activeMigrations)
+	notForeverToleratedVMIs := c.filterNotToleratedVMIs(now, notMigratingVMIs, node.Spec.Taints)
+
+	// Ensure we enqueue VMIs again which are only temporary tolerated
+	// and treat the others as migration candidates
+	migrationCandidates := []*notolerate{}
+	for _, candidate := range notForeverToleratedVMIs {
+		if len(candidate.NotTolerated) > 0 {
+			migrationCandidates = append(migrationCandidates, candidate)
+		} else if candidate.FirstReque != nil {
+			c.Queue.AddAfter(node.Name, candidate.FirstReque.Sub(now))
+		}
+	}
+
+	diff := int(math.Min(float64(5-len(activeMigrations)), float64(len(migrationCandidates))))
```

Comment:
would probably be better to have a constant for this 5 :)

Reply:
done